trie/triedb/pathdb/history.go

```diff
@@ -0,0 +1,594 @@
+// - account indexes/storage indexer buffer is not aligned
+//
+// note, these situations are allowed:
+//
+// - empty account data: all accounts were not present
+// - empty storage set: no slots are modified
+func (r *decoder) verify() error {
+	if len(r.accountIndexes)%accountIndexSize != 0 || len(r.accountIndexes) == 0 {
+		return fmt.Errorf("invalid account index, len: %d", len(r.accountIndexes))
+	}
+	if len(r.storageIndexes)%storageIndexSize != 0 {
+		return fmt.Errorf("invalid storage index, len: %d", len(r.storageIndexes))
+	}
+	return nil
+}
+
+// readAccount parses the account from the byte stream with specified position.
+func (r *decoder) readAccount(pos int) (accountIndex, []byte, error) {
+	// Decode account index from the index byte stream.
+	var index accountIndex
+	if (pos+1)*accountIndexSize > len(r.accountIndexes) {
+		return accountIndex{}, nil, errors.New("account data buffer is corrupted")
+	}
+	index.decode(r.accountIndexes[pos*accountIndexSize : (pos+1)*accountIndexSize])
+
+	// Perform validation before parsing account data, ensure
+	// - account is sorted in order in byte stream
+	// - account data is strictly encoded with no gap inside
+	// - account data is not out-of-slice
+	if r.lastAccount != (common.Hash{}) {
+		if bytes.Compare(r.lastAccount.Bytes(), index.Hash.Bytes()) >= 0 {
+			return accountIndex{}, nil, errors.New("account is not in order")
+		}
+	}
+	if index.Offset != r.lastAccountRead {
+		return accountIndex{}, nil, errors.New("account data buffer is gaped")
+	}
+	last := index.Offset + uint32(index.Length)
+	if uint32(len(r.accountData)) < last {
+		return accountIndex{}, nil, errors.New("account data buffer is corrupted")
+	}
+	data := r.accountData[index.Offset:last]
+
+	r.lastAccount = index.Hash
+	r.lastAccountRead = last
+
+	return index, data, nil
+}
+
+// readStorage parses the storages from the byte stream with specified account.
+func (r *decoder) readStorage(aIndex accountIndex) ([]common.Hash, map[common.Hash][]byte, error) {
+	var (
+		last    common.Hash
+		list    []common.Hash
+		storage = make(map[common.Hash][]byte)
+	)
+	for j := 0; j < int(aIndex.StorageNumber); j++ {
+		var (
+			index slotIndex
+			start = (aIndex.StorageOffset + uint32(j)) * uint32(storageIndexSize)
+			end   = (aIndex.StorageOffset + uint32(j+1)) * uint32(storageIndexSize)
+		)
+		// Perform validation before parsing storage slot data, ensure
+		// - slot index is not out-of-slice
+		// - slot data is not out-of-slice
+		// - slot is sorted in order in byte stream
+		// - slot indexes is strictly encoded with no gap inside
+		// - slot data is strictly encoded with no gap inside
+		if start != r.lastSIndexRead {
+			return nil, nil, errors.New("storage index buffer is gapped")
+		}
+		if uint32(len(r.storageIndexes)) < end {
+			return nil, nil, errors.New("storage index buffer is corrupted")
+		}
+		index.decode(r.storageIndexes[start:end])
+
+		if last != (common.Hash{}) {
```

Comment:
Interestingly, you don't need this check because last will be smaller than anything being the zero hash, since you can't set the zero hash slot anyway. It's ok like this too, probably cleaner, just an interesting fact.